### PR TITLE
[POC] Include header and summary rows in cell navigation 

### PR DESCRIPTION
**Pull request by amanmahajan7**
https://www.w3.org/TR/wai-aria-practices-1.1/#keyboard-interaction-for-data-grids

In general all the rows/cell should be reachable via keyboard. I checked a few third party grids and they support it as well
https://material-ui.com/components/data-grid/
https://www.ag-grid.com/example.php?gclid=Cj0KCQjwpf2IBhDkARIsAGVo0D1BfZ-ISqPQGf6qG7ngaVzyxzUpeOUJnp5PlEdez971H8Pan7u_rFMaApaBEALw_wcB

### CODE REVIEW

File: src/Columns.tsx [hunk 1]
```diff
@@ -51,9 +51,11 @@ export const SelectColumn: Column<any, any> = {
     return (
       <SelectCellFormatter
         aria-label="Select All"
-        isCellSelected={false}
+        isCellSelected={props.isCellSelected}
```
Comment (amanmahajan7) [inline src/Columns.tsx:54]: Now all the cells including header and summary receive `isCellSelected` prop

File: src/DataGrid.tsx [hunk 2]
```diff
@@ -329,15 +331,19 @@ function DataGrid<R, SR, K extends Key>(
 
   const hasGroups = groupBy.length > 0 && typeof rowGrouper === 'function';
   const minColIdx = hasGroups ? -1 : 0;
-  const selectedCellIsWithinBounds = isCellWithinBounds(selectedPosition);
+  const maxColIdx = columns.length - 1;
+  const minRowIdx = -1; // change in to 0?
```
Comment (amanmahajan7) [inline src/DataGrid.tsx:335]: Had to set it to -1 so we can continue to use `rows[index]`

File: src/DataGrid.tsx [hunk 3]
```diff
@@ -703,18 +733,24 @@ function DataGrid<R, SR, K extends Key>(
...
       let right = left + width;
-      const row = rows[rowIdx];
-      if (!isGroupRow(row)) {
-        const colSpan = getColSpan(column, lastFrozenColumnIndex, { type: 'ROW', row });
-        if (colSpan !== undefined) {
-          const { left, width } = columnMetrics.get(columns[column.idx + colSpan - 1])!;
-          right = left + width;
-        }
+
+      const colSpan = getSelectedCellColSpan({
```
Comment (amanmahajan7) [inline src/DataGrid.tsx:742]: Need to add a test for header and summary navigation when colSpan is defined

File: src/DataGrid.tsx [hunk 4]
```diff
@@ -779,18 +815,18 @@ function DataGrid<R, SR, K extends Key>(
...
+        return ctrlKey ? { idx: 0, rowIdx: minRowIdx } : { idx: 0, rowIdx };
       case 'End':
         // If row is selected then move focus to the last row.
         if (isRowSelected) return { idx, rowIdx: rows.length - 1 };
-        return ctrlKey
-          ? { idx: columns.length - 1, rowIdx: rows.length - 1 }
-          : { idx: columns.length - 1, rowIdx };
+        return ctrlKey ? { idx: maxColIdx, rowIdx: maxRowIdx } : { idx: maxColIdx, rowIdx };
       case 'PageUp': {
+        if (selectedPosition.rowIdx === minRowIdx) return selectedPosition;
```
Comment (amanmahajan7) [inline src/DataGrid.tsx:824]: `PageUP/Down` does not move the selected cell to header and summary row. Not sure if we need to do that

File: src/SummaryRow.tsx [hunk 5]
```diff
@@ -1,18 +1,23 @@
...
+import type { CalculatedColumn, RowRendererProps } from './types';
+import { useRovingRowRef } from './hooks';
 
 type SharedRowRendererProps<R, SR> = Pick<RowRendererProps<R, SR>, 'viewportColumns' | 'rowIdx'>;
 
-interface SummaryRowProps<R, SR> extends SharedRowRendererProps<R, SR> {
-  'aria-rowindex': number;
+interface SummaryRowProps<R, SR>
+  extends SharedRowRendererProps<R, SR>,
+    Pick<React.HTMLAttributes<HTMLDivElement>, 'aria-rowindex'> {
```
Comment (nstepien) [inline src/SummaryRow.tsx:15]: `aria-rowindex` is optional now, is this intentional?

File: src/hooks/useRovingRowRef.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+const rowSelected = css`
+  outline: none;
+
+  &::after {
+    content: '';
+    position: absolute;
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
```
Comment (nstepien) [inline src/hooks/useRovingRowRef.ts:15]: Should check other cases where we can simplify this.
```suggestion
    inset: 0;
```

**Review by nstepien — COMMENTED** [review]
Not done reviewing

File: src/SummaryRow.tsx (hunk 5 above)
Comment (amanmahajan7) [reply src/SummaryRow.tsx]: It was unintentional, reverted it

File: src/DataGrid.tsx [hunk 7]
```diff
@@ -473,30 +489,29 @@ function DataGrid<R, SR, K extends Key>(
   }
 
   function onGridFocus() {
-    if (selectedCellIsWithinBounds) return;
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/DataGrid.tsx:476]:
```
{!selectedCellIsWithinBounds && (
    <div className={focusSinkClassname} tabIndex={0} onFocus={onGridFocus} />
)}
```

We do not render `focusSink` if the cell is within selection bounds

File: test/copyPaste.test.tsx [hunk 8]
```diff
@@ -111,3 +125,20 @@ test('should cancel copy/paste on escape', () => {
...
 });
+
+test('should not allow paste on header or summary cells', () => {
+  setup();
+  userEvent.click(getCellsAtRowIndex(0)[0]);
+  copySelectedCell();
+  expect(getSelectedCell()).toHaveClass(copyCellClassName);
+  userEvent.keyboard('{arrowup}');
+  pasteSelectedCell();
+  expect(getSelectedCell()).not.toHaveClass(copyCellClassName);
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply test/copyPaste.test.tsx]: Yeah, I have updated copy/paste tests

File: src/utils/selectedCellUtils.ts [hunk 9]
```diff
@@ -32,36 +32,87 @@ interface GetNextSelectedCellPositionOpts<R, SR> {
...
+    if (!isGroupRow(row)) {
+      return getColSpan(column, lastFrozenColumnIndex, { type: 'ROW', row });
+    }
+    return undefined;
+  }
+
+  if (summaryRows) {
+    return getColSpan(column, lastFrozenColumnIndex, {
+      type: 'SUMMARY',
+      row: summaryRows[rowIdx]
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/utils/selectedCellUtils.ts]: Good catch. Fixed now

**Pull request merged**